프로젝트 관리 앱 [STEP2-2] 두기

doogie97 opened the pull request:
# 프로젝트 관리 앱 pr

@ ttozzi
안녕하세요 또치!
미리 말씀드리긴 했지만 예정일보다 한참을 늦게 보내서 죄송합니다...!

저장소를 제외한 기능 구현 완료하여 pr 보냅니다!

### MockStorage
![](https://i.imgur.com/adnJzRz.png)
아직은 로컬 및 원격 저장소를 구현하기 전이기에 Storageable이라는 프로토콜을 만들어 해당 프로토콜을 준수하는 객체가 저장소가 될 수 있도록 구현하였습니다

### 저장소 공유
MainViewController에서는 read, delete를 DetailViewController에서는 create, update를 하기에 저장소는 공유가 되어야 하는데 그 방법을 총 세가지를 생각했었습니다

1. detailview를 present할 때 MainViewModel을 주입시킨다
2. storage를 싱글톤으로 구현한다
3. DI Container를 만들어 SceneDelegate에서 MainViewcontroller에 storage를 주입해서 단계별로 주입을 한다

1번같은경우는 detailview또한 viewmodel을 만들어 역할 분리를 하는 것이 맞다고 생각해 제외
2번같은 경우는 저장소를 싱글톤으로 공유하게 된다면 의도치 않은 곳에서 접근할 가능성이 있기에 제외하였으며 3번의 방법을 선택 하였습니다
(물론 이번에 처음 알게된 개념이라 완전한 구현은 어려워 개념은 가져가고 storage주입을 목적으로 약간 따라하는 식으로만 구현해 보았습니다)

![](https://i.imgur.com/rz7mZOU.png)


### input/ output

![](https://i.imgur.com/9mieDDw.png)

input은 viewmodel이 이벤트를 받아 해야할 행동
output은 viewmodel이 view로 전달하는 데이터 
정도로 이해를 했는데 알맞게 구현한 것이 맞을까요?


### cell의 index만 받아 viewModel에서 해당 list를 반환 하기

![](https://i.imgur.com/xFZWyDr.png)

1. 위와 같은 로직을 통해 cell을 반환하는 기능을 구현해 보았는데 completion으로 이렇게 반환하는게 또치가 원하신 모습과 얼마나 차이가 있을지 여쭤보고 싶습니다

2. 그리고 각각의 output이 driver로 구현되어 있는데 driver는 원래 filter나 map 사용이 안되는건지 여쭤보고 싶습니다(일단 반응을 안해 drive한 뒤에 onNext로 넘겨주도록 구현을 했습니다)

3. drive를 하게 되면 무한 반복되는 현상이 있어서 원하는 일을 한 뒤에 바로 dispose를 해주도록 했는데 이렇게 된다면 문제가 있을까요?(제가 예상하기로는 onNext이후 작업이 오래걸린다면 작업이 끝나기 전에 dispose될 것 같다 라는 생각을 하긴 했는데 이걸 확인 할 방법이 없어 여쭤보게 되었습니다!)

doogie97 committed:
♻️refactor : tableView 생성 관련 중복 코드 하나의 메서드로 통합

doogie97 committed:
♻️refactor : view 교체 시점 변경

doogie97 committed:
♻️refactor : view에 요소 표시할 때 bind -> drive로 변경

doogie97 committed:
💎feat : MockStorage구현 및 output 구현

doogie97 committed:
💎feat : input/output protocol구현 및 peekList 메서드 구현

doogie97 committed:
💎feat : Storege 프로토콜 구현 메서드 구현

doogie97 committed:
💎feat : delete기능 구현

doogie97 committed:
💎feat : detailView생성시 viewModel을 추가로 주입받도록 수정

+ 주입받은 파라미터를 전역변수로 생성해 view구성 또한 viewDidLoad에서 할 수 있도록 수정

doogie97 committed:
♻️refactor : creat, update시 List가 날짜순으로 정렬되도록 수정

doogie97 committed:
🔨fix : 무한으로 셀이 터치되는 현상 수정

doogie97 committed:
♻️refactor : 접근제어 설정 변경

doogie97 committed:
💎feat : create 기능 구현

doogie97 committed:
💎feat : Update기능 구현 및 메서드 분리

doogie97 committed:
💎feat : detail view 진입 경로에 따라 버튼 및 ui수정 가능 여부 설정

doogie97 committed:
♻️refactor : headerView countLabel 수정

doogie97 committed:
♻️refactor : DetailView shadow설정

doogie97 committed:
♻️refactor : cell 간격 및 bottm line 설정

doogie97 committed:
⚰️del : 쓸모없는 코드 삭제

doogie97 committed:
♻️refactor : in,out을 하나로 묶어 MainViewModel에서 채택

doogie97 committed:
✏️chore : read역할을 하는 list를 Storage프로토콜에 추가

doogie97 committed:
💎feat : DI Container로 mainVC에 storage,viewmodel 주입

doogie97 committed:
⚰️del : 쓸모없는 코드 삭제

doogie97 committed:
♻️refactor : DetailViewController를 add, edit 두개로 분리

doogie97 committed:
♻️refactor : Storege -> Storegeable로 네이밍 수정

doogie97 committed:
♻️refactor : container singletone으로 변경

add나 edit VC init시 의존성 주입 방법 문제로 일단 singletone으로 변경

doogie97 committed:
💎feat : DetailViewModel생성

doogie97 committed:
⚰️del : 쓸모없는 코드 삭제

doogie97 committed:
💎feat : AddVC생성 메서드 구현

doogie97 committed:
♻️refactor : Container를 통해 AddVC 초기화로 방식 변경

doogie97 committed:
🔨fix : List Edit이 불가하던 현상 수정

doogie97 committed:
💎feat : EditVC생성 메서드 구현

doogie97 committed:
♻️refactor : Container를 통해 EditVC 초기화로 방식 변경

doogie97 committed:
✏️chore : 폴더 구조 변경

doogie97 committed:
⚰️del : 쓸모없는 코드 삭제

doogie97 committed:
✏️chore :  RxGesture 추가

doogie97 committed:
♻️refactor : changeListType의 파라미터 수정

doogie97 committed:
💎feat : ListType에 따라 alertTitle을 반환하는 연산 프로퍼티 생성

doogie97 committed:
♻️refactor : list추출 기능을 peekList메서드로 교체

doogie97 committed:
♻️refactor : changeListType의 파라미터 추가

doogie97 committed:
💎feat : changeListType 메서드 구현

doogie97 committed:
💎feat : ListType에 Direction 추가

doogie97 committed:
💎feat : 셀 longPress 할 경우 이동 기능 구현

doogie97 committed:
♻️refactor :container를 singletone -> self를 주입하도록 변경

doogie97 committed:
♻️refactor : data sort시점 변경

doogie97 committed:
🔨fix : text가 textView밖으로 나가는 현상 수정

doogie97 committed:
💎feat : modalPresentationStyle 설정

doogie97 committed:
♻️refactor : edit모드로 변경시 toggle을 이용하도록 수정

doogie97 committed:
📝docs : README Update

TTOzzi commented on ProjectManager/Podfile:
```diff
@@ -8,6 +8,7 @@ pod 'SwiftLint'
 pod 'RxSwift'
 pod 'RxCocoa'
 pod 'SnapKit'
+pod "RxGesture"
```
정말정말 사소한건데 쌍따옴표 눈에 띄어서 코멘트 남겨요 +_+

TTOzzi commented on ProjectManager/ProjectManager/MockStorage.swift:
```diff
@@ -0,0 +1,40 @@
+    func deleteList(listItem: ListItem)
+}
+
+class MockStorage: Storegeable {
```
`final` 붙여주면 좋을 것 같아요!

TTOzzi commented on ProjectManager/ProjectManager/MockStorage.swift:
```diff
@@ -0,0 +1,40 @@
+        ListItem(title: "안녕 똥이야!", body: "모 \n 어쩔건데!! \n ㅋㄷㅋㄷ", deadline: Date(timeIntervalSince1970: 751231235), type: .done)
+    ]
+    
+    lazy var list = BehaviorRelay<[ListItem]>(value: dummyList)
```
`lazy var` 로 선언하신 이유가 있을까요? 👀

doogie97 replied:
dummyList를 생성할때 바로 넣어주고 있어서 lazy var로 선언했는데 
![image](https://user-images.githubusercontent.com/82325822/179337347-64e5e9ce-5f09-4237-8813-2a1aa68f4890.png)

이런식으로 수정하는 걸 의도 하신걸까요??(아직 수정은 하지 않았습니다!)

TTOzzi commented on ProjectManager/ProjectManager/MockStorage.swift:
```diff
@@ -0,0 +1,40 @@
+    }
+    
+    func updateList(listItem: ListItem) {
+        list.accept(list.value.filter { listItem.id != $0.id } + [listItem])
```
음 특정 `id`가 아닌 `ListItem`을 찾기 위해서 배열 전체를 순회하기 보단, 모든 `id`가 유니크하단 가정 하에 특정 `id`인 것을 찾는 게 더 효율적이지 않을까요? 🤔 
수정 작업에 해당하는`updateList` 메서드가 불렸는데, 수정할 `id`가 같은 `ListItem`이 없는 경우에 대한 예외처리도 가능할 것 같구요.

추가로 요건 개인적인 생각이 살짝 첨가된 코멘트인데, 한 줄로 깔끔하게 작성하는 것도 좋지만 여러 줄로 표현된 코드가 더 이해하기 쉬울 때도 있습니다.
`list.value.filter { listItem.id != $0.id }` 요정돈 메서드 내부에 `let`으로 선언해서 의미를 부여해주면 더 이해하기 쉬운 코드가 될 것 같아요.

doogie97 replied:
```swift
 func updateList(listItem: ListItem) {
        delete(listItem)
        
        list.accept(listArray + [listItem])
    }
    
    func deleteList(listItem: ListItem) {
        delete(listItem)
        list.accept(listArray)
    }
    
    private func delete(_ listItem: ListItem) {
        let index = listArray.firstIndex {
            $0.id == listItem.id
        }
        
        guard let index = index else {
            return
        }

        listArray.remove(at: index)
    }
```
위와 같이 수정해보았습니다

TTOzzi commented on ProjectManager/ProjectManager/MockStorage.swift:
```diff
@@ -0,0 +1,40 @@
+    }
+    
+    func deleteList(listItem: ListItem) {
+        list.accept(list.value.filter { listItem.id != $0.id })
```
요기도 마찬가지로 배열 전체를 확인할 필요 없이, `id`가 같은 `ListItem`이 나올 때 까지만 확인하면 되지 않을까요? 👀

TTOzzi commented on ProjectManager/ProjectManager/Model/ListItem.swift:
```diff
@@ -30,4 +30,48 @@ enum ListType {
         }
     }
+    
+    var firstAlertTitle: String {
+        switch self {
+        case .todo:
+            return "Move to Doing"
+        case .doing:
+            return "Move to Todo"
+        case .done:
+            return "Move to Todo"
+        }
+    }
+    
+    var firstDirection: ListType {
+        switch self {
+        case .todo:
+            return .doing
+        case .doing:
+            return .todo
+        case .done:
+            return .todo
+        }
+    }
```
반환값
title - direction 쌍이 second ~~ 에도 반복되는데, 별도의 데이터 타입으로 묶어서 표현해줄 수도 있을 것 같아요 🤔 
데이터 타입을 하나 더 만드는 게 과하다면, title은 항상 반환값 `ListType`과 매칭되니 title만 묶어줘도 깔끔할 것 같네요!

doogie97 replied:
```swift
    struct Direction {
        let title: String
        let type: ListType
    }
    
    var firstDirection: Direction {
        switch self {
        case .todo:
            return Direction(title: "Move to Doing", type: .doing)
        case .doing:
            return Direction(title: "Move to Todo", type: .todo)
        case .done:
            return Direction(title: "Move to Todo", type: .todo)
        }
    }
    
    var secondDirection: Direction {
```

위와 같이 Direction타입을 생성해 반환할 수 있도록 해보았습니다
(아래 second도 동일해 코멘트에서는 생략하였습니다)

TTOzzi commented on ProjectManager/ProjectManager/View/Detail/AddViewController.swift:
```diff
@@ -0,0 +1,58 @@
+        let listItem = ListItem(title: detailView.titleTextField.text ?? "",
+                                body: detailView.bodyTextView.text ?? "",
+                                deadline: detailView.deadlinePicker.date)
+        viewModel.creatList(listItem: listItem)
```
`doneButton` 버튼이 눌렸을 때, 뷰의 데이터를 조립하기보단 뷰는 데이터를 그대로 뷰모델로 넘기기만 하고 뷰모델에서 데이터 조립, 값이 없을 때의 기본값 할당을 수행하는 게 좋을 것 같아요.
`detailView`에서 발생하는 사용자 이벤트를 뷰모델로 전달하도록 바인딩한 후 title/body 입력, datePicker 값이 변경될 때 마다 뷰모델의 상태값을 업데이트 하고, done 버튼이 눌렸을 땐 뷰에서 값을 전달하는 게 아닌 이미 뷰에 표시되고 있는 값을 뷰모델이 갖고 있을테니 그대로 저장 로직을 수행하도록요!
뷰모델의 input 이름도 "리스트를 만들어줘!" 보단, 최대한 디테일한 정보를 가지지 않는 뷰의 관점에서 "완료 버튼이 눌렸어!"가 되어야 할 것 같구요 🤔

doogie97 replied:
```swift
final class AddViewModel: DetailViewModelable {
    private let storage: Storegeable
    var list: ListItem
    
    init(storage: Storegeable, list: ListItem?) {
        self.storage = storage
        self.list = list ?? ListItem(title: "", body: "", deadline: Date())
    }
    
    func changeTitle(_ text: String?) {
        list.title = text ?? ""
    }
    
    func changeDaedLine(_ date: Date?) {
        list.deadline = date ?? Date()
    }
    
    func changeBody(_ text: String?) {
        list.body = text ?? ""
    }
    
    func touchDoneButton() {
        storage.creatList(listItem: list)
    }
}
```
위와 같이 구성해 뷰의 요소가 바뀔 때 마다 viewModel이 가지고 있는 list가 변경될 수 있도록 구현하였습니다

그리고 말씀하신 대로 update해줘, create해줘가 아닌 button이 눌렸어! 라고 구현하기 위하여 add와 edit의 뷰모델도 분리해주었습니다

TTOzzi commented on ProjectManager/ProjectManager/View/Detail/DetailView.swift:
```diff
@@ -54,43 +54,53 @@ final class DetailView: UIView {
-            naviItem.title = ListType.todo.title
-            return
-        }
+    func setAddView() {
+        naviItem.title = ListType.todo.title
+    }
```
요건 외부에서 호출해줘야 하는 이유가 있을까요? 
`DetailView`가 생성될 때 설정해줘도 될 것 같아서요 🤔

doogie97 replied:
원래 addVC는 list를 갖고 있지 않아 뷰구성 로직이 달라졌는데 지금은 list를 갖고 있어 같은 로직으로도 뷰구성이 가능해 삭제해줬습니다!

TTOzzi commented on ProjectManager/ProjectManager/View/Detail/DetailView.swift:
```diff
@@ -54,43 +54,53 @@ final class DetailView: UIView {
+    }
+    
+    func changeEditable() {
+        titleTextField.isEnabled.toggle()
+        deadlinePicker.isUserInteractionEnabled.toggle()
+        bodyTextView.isEditable.toggle()
```
인자를 전달해 명확히 값을 설정하는 게 아니라 `toggle`로 값을 조절한다면, 원하는 시점에 뷰가 편집이 가능한 상태인 것을 보장할 수 있을까요? 
뷰가 편집 가능한 상태인가를 뷰모델의 상태값으로 관리하고, 그 값과 위 세 뷰의 enabled 값들을 바인딩해주면 현재 뷰가 편집 가능 여부를 명확히 통제 할 수 있는 구조가 될 것 같습니다.

doogie97 replied:
```swift
        viewModel.isEditable.bind(onNext: { [weak self] in
            self?.detailView.titleTextField.isEnabled = $0
            self?.detailView.deadlinePicker.isUserInteractionEnabled = $0
            self?.detailView.bodyTextView.isEditable = $0
        })
        .disposed(by: disposebag)
```

위와 같은 로직을 통해 현재 상태의 반대 값이 아닌 뷰모델의 편집 가능 여부를 기반으로 뷰에 반영할 수 있도록 하였습니다

TTOzzi commented on ProjectManager/ProjectManager/View/Detail/EditlViewController.swift:
```diff
@@ -0,0 +1,70 @@
+
+final class EditlViewController: UIViewController {
+    private let viewModel: DetailViewModel
+    private var listItem: ListItem
```
음 뷰모델이 존재하는데, `listItem`이 따로 정의되어 있어야 할 이유가 있을까요?
뷰모델은 뷰에서 발생하는 이벤트를 받아 전달하는 것 뿐만 아니라, 뷰가 표시해야 할 상태값을 관리하고 뷰로 전달하는 역할도 합니다.
뷰는 뷰모델에게서 값을 받아 업데이트하도록 처리하고 `listItem`은 뷰모델 안으로 넣어줘야 할 것 같아요.
지금은 데이터의 원천이 두 곳으로 나뉜 것 같습니다 🤔

TTOzzi commented on ProjectManager/ProjectManager/View/Detail/EditlViewController.swift:
```diff
@@ -0,0 +1,70 @@
+        listItem.body = detailView.bodyTextView.text ?? ""
+        listItem.deadline = detailView.deadlinePicker.date
+        
+        viewModel.updateList(listItem: listItem)
```
여기도 `AddViewController`와 마찬가지로 데이터를 조립하는 로직은 뷰모델단으로 옮겨주면 좋을 것 같아요.
`listItem`이 뷰모델 내부로 이동하면, 이 수정 로직도 뷰모델에서 수행할 수 있겠죠?

TTOzzi commented on ProjectManager/ProjectManager/View/Detail/EditlViewController.swift:
```diff
@@ -0,0 +1,70 @@
+            self.dismiss(animated: true)
+        }
+        detailView.leftButton.title = "Cancel"
+        detailView.changeEditable()
```
`detailView.leftButton.title`이 `"Cancel"` 인가 로 로직을 분기하는 건 좋은 로직은 아닌 것 같습니다.
`detailView.leftButton.title`의 값이 변경되면 이곳에도 수정이 필요하고, 코드를 처음 보는 사람에겐 "`detailView.leftButton.title`이 `"Cancel"` 인가"가 무엇을 의미하는지 파악하기도 어려울 거예요.
뷰에 로직이 모두 구현되어 있다 보니 테스트도 어려운 구조가 되어버렸구요.

뷰모델에서 현재 화면의 상태(편집 모드인지?)를 갖고 있고, 좌측 버튼을 누르는 이벤트가 발생하면 뷰모델로 전달한 후 뷰모델 내에서 현재 상태에 따라 로직을 분기하고, 그에 따른 결과를 뷰로 전달해 뷰를 업데이트 하는 흐름이 되어야 할 것 같습니다.

예를 들면,
`leftButton` 클릭 
-> 뷰모델로 `tapLeftButton`input 전달 
-> 뷰모델에선 뷰가 현재 편집 모드인지 확인(로직 수행) 
-> 현재 편집 모드가 아니라면 `편집 모드로 변경` output 전달 / 편집 모드라면 `dismiss` output 전달 
-> 뷰에선 output을 받아 뷰를 업데이트
의 흐름이 될 것 같네요.

doogie97 replied:
```swift
var isEditable = BehaviorRelay<Bool>(value: false)

    func touchLeftButton(_ vc: EditViewController) -> String? {
        if isEditable.value {
            vc.dismiss(animated: true)
            return nil
        } else {
            self.isEditable.accept(true)
            return "Cancel"
        }
    }
```
viewmodel에서는 위 로직을 구현해주었고


```swift
        detailView.leftButton.rx.tap
            .bind(onNext: { [weak self] in
                guard let self = self else {
                    return
                }
                let title = self.viewModel.touchLeftButton(self)
                self.detailView.leftButton.title = title
            })
            .disposed(by: disposebag)
```
뷰컨트롤러에서는 위로직을 통해 반영해주었습니다

다만 여기서 touchLeftButton이 string을 반환한다는것이 좀 마음에 걸리는데(처음 보는 사람은 touchLeftButton이 title을 전달하는 것이라고 생각하지 못할 것이니) 더 나은 방법이 있을까요??

TTOzzi replied:
디엠으로도 말씀드렸듯이 뷰모델 구현에선 인자로 뷰컨을 받아 뷰컨의 메서드를 호출하는 게 아닌,
```swift
var dismiss = PublishRelay<Void>()
```
와 같이 output을 정의하고, `dismiss.accept(())` 의 형태로 뷰에 output 전달 -> 뷰에선 output을 받았을 때 dismiss 하는 형태가 되어야 할 것 같아요.
반환값인 `String?` 또한 뷰모델의 input에서 바로 반환하는 게 아니라, `isEditable` output을 엮어주는 뷰컨에서 `isEditable`이 true 이면 "Cancel"로 변경하도록 구현해야 할 것 같구요.

항상 뷰에서 사용자 이벤트 발생 -> input 함수 호출로 전달 -> 뷰모델 내부에서 상태 변경(로직 수행) -> output 으로 값 방출 -> 뷰에서 받아 뷰 업데이트 의 흐름을 지켜주세요 🙇‍♂️

TTOzzi commented on ProjectManager/ProjectManager/View/Main/ListTableViewCell.swift:
```diff
@@ -16,7 +16,12 @@ final class ListTableViewCell: UITableViewCell {
         fatalError("init(coder:) has not been implemented")
     }
-    
+    private lazy var viewview: UIView = {
```
뷰뷰는 오타..일까요? 👀

TTOzzi commented on ProjectManager/ProjectManager/View/Main/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+    
+    private lazy var todoTableView = listTableView()
+    private lazy var  doingTableView = listTableView()
+    private lazy var  doneTableView = listTableView()
```
여기 두 친구는 거리가 좀 먼 것 같아요 +_+

TTOzzi commented on ProjectManager/ProjectManager/View/Main/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+        setInitailView()
+    }
+    
+    private func setInitailView() {
```
Initail 오타인 것 같아요 +_+

TTOzzi commented on ProjectManager/ProjectManager/View/Main/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+    
+    @objc private func didTapAddButton() {
+        self.present(container.makeAddViewController(), animated: true)
+    }
```
현재 로직상으론 꼭 필요한 부분은 아니지만, 가능하면 여기도 뷰모델을 거치도록 구현해보면 좋을 것 같아요.
뷰모델에 추가 버튼이 눌렸음을 전달하고, 뷰모델에선 `AddViewController`를 띄워 달라는 output을 뷰로 전달해 업데이트하도록요.
미리 구조를 짜두면, 나중에 추가 버튼이 눌렸을 때 요구사항이 추가되거나 변경됐을 때 좀 더 쉽게 수정이 가능합니다.

doogie97 replied:
위 사진 부분 말고도 다른 화면전환하는 모든 부분들다 포함해서 화면 전화또한 뷰컨에서 하기보다는 뷰모델에서 하는게 맞는걸까요??
```swift
struct MainViewModel{
    func touchAddbutton(_ vc: MainViewController) {
        vc.present(AddViewController())
    }
}
```
이런 식의 형태가 될것같고 이 메서드를 뷰컨에서 호출해주는?

TTOzzi replied:
위에 남긴 `detailView.leftButton` 관련 코멘트와 같습니다!
뷰컨을 뷰모델로 전달하는 게 아닌, input을 넣고 output을 받아 뷰에서 처리하도록 구현해주세요.

추후 더 나은 구조로 개선한다면, 화면 이동을 담당하는 객체를 따로 만들어 관리할 수도 있습니다.

TTOzzi commented on ProjectManager/ProjectManager/View/Main/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+    
+    private func didtapCell(_ listItem: ListItem) {
+        self.present(container.makeEditViewController(listItem), animated: true)
+    }
```
여긴 버튼이 바로 눌리는 게 아닌 뷰모델의 output을 받아 호출될테니 호출하는 시점에서 `didtapCell` 보단 `EditViewController`를 띄워준다는 것을 표현할 수 있는 네이밍이 더 좋을 것 같아요.

TTOzzi commented on ProjectManager/ProjectManager/View/Main/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+            .disposed(by: disposebag)
+        
+        list
+          .map { "\($0.count)"}
```
이 부분도 어찌보면 뷰에 표시하기위한 데이터로의 가공 로직이니 뷰모델에서 미리 스트림을 나눠서 반환하고 뷰에선 가공 없이 이어주기만 하면 더 좋을 것 같아요.

TTOzzi commented on ProjectManager/ProjectManager/View/Main/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+                
+                self.viewModel.peekList(index: index.row, type: type) {
+                    self.didtapCell($0)
+                }
```
흠.. 뷰모델로 input을 전달하는 흐름과 뷰모델에서 output 을 받아 뷰를 업데이트하는 흐름이 명확하게 나눠지지 못한 것 같습니다.
```suggestion
                self.viewModel.peekList(index: index.row, type: type)
```
여기선 input 전달까지만 수행하고, 다른 곳에서 뷰모델의 output을 받아 뷰를 업데이트하는 로직을 구현해야 합니다.
output의 전달은 함수 내부에서 로직을 수행한 뒤 지금처럼 함수 인자로 클로저를 전달하는 게 아니라, 뷰모델에 output 전달 통로를 정의하고 그곳을 통해 결과값을 전달하는 형태가 되어야 할 것 같아요.

`MainViewModel`에 `ListItem`을 방출하는 `presentEditView` output을 정의하고, 뷰모델에서 `peekList`가 호출됐을 때 로직을 수행한 뒤 `presentEditView`를 통해 `ListItem`을 방출하면 뷰에선 `presentEditView`가 방출한 값을 받아 `EditViewController`를 띄워주도록요.

TTOzzi replied:
(`peek`은 데이터 구조에서 선택 보단 다른 의미로 쓰이는 경우가 많아 select 와 같은 명확한 이름으로 변경하는 게 좋을 것 같아요 +_+)

doogie97 replied:
이 부분에 대해서 고민을 해봤을 때 굳이 item을 직접 넘겨주기 보다 몇 번째 셀이 선택됐는지, 그 셀은 어떤 타입의 item인지만 editViewModel로 전달해줘서 그 곳에서 받은 정보로 storage에서 item을 뽑아 오는게 좋을 것 같다는 생각이 들어 말씀드린 방식으로 수정하였습니다

TTOzzi commented on ProjectManager/ProjectManager/View/Main/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+        
+        tableView.rx.itemDeleted
+            .bind(onNext: {
+                self.viewModel.deleteList(index: $0.row, type: type)
```
여긴 변경된 list 를 방출하는 output이 이미 구현되어있어 흐름이 분리되었군요!

doogie97 replied:
이 부분도 제가 잘 이해하지 못했습니다
조금 더 상세하게 설명해 주시면 감사하겠습니다

TTOzzi replied:
아! 요건 `deleteList` input 으로 raw 한 값 전달 -> 뷰모델 내에서 로직 처리 -> 구현되어있던 list output으로 뷰에 업데이트된 내용 반영 흐름으로 잘 구현하셨단 코멘트입니다!

TTOzzi commented on ProjectManager/ProjectManager/View/Main/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+        
+        let alert = makeAlert(index: indexPath.row, type: type)
+
+        guard let popovecon = alert.popoverPresentationController else {
```
`popovecon` 도 오타..? 같습니다.

TTOzzi commented on ProjectManager/ProjectManager/View/Main/MainViewController.swift:
```diff
@@ -0,0 +1,205 @@
+
+        popovecon.sourceView = cell
+        
+        let checkLoation = location.y < UIScreen.main.bounds.height * 3 / 5
```
`checkLoation`보단 is~~ 형태로 해당 Bool 값이 어떤 의미를 가지는지 표현해주면 더 좋을 것 같아요.
현재 코드론 `location.y < UIScreen.main.bounds.height * 3 / 5` 조건이 무엇을 의미하는 지 이해하기 어렵네요 🤔

doogie97 replied:
list 타입을 바꿀때 나오는 action sheet가 화면의 3/5지점에서 부터는 위로 나올 수 있게 하는 로직인데요 뭐라고 네이밍을 지으면 좋을까요...??

TTOzzi replied:
아하 그런 의미였군요!
저도 영어를 잘 못해서... 아래쪽 방향으로 보여야 하는지를 분기하는 값인듯하니, `shouldShowBelow`는 어떨까요? 👀

TTOzzi commented on ProjectManager/ProjectManager/ViewModel/MainViewModel.swift:
```diff
@@ -7,43 +7,88 @@
+    var doneList: Driver<[ListItem]> { get }
+}
+
+protocol MainViewModelInput {
+    func isOverDeadline(listItem: ListItem) -> Bool
+    func peekList(index: Int, type: ListType, completion: @escaping ((ListItem) -> Void))
+    func deleteList(index: Int, type: ListType)
+    func changeListType(index: Int, type: ListType, to: ListType)
+}
```
뷰모델에서 input은 뷰로부터 발생한 이벤트를 의미한다고 생각하는데요.
`isOverDeadline`이 뷰에서 발생한 이벤트일까요? 🤔 
그리고 input에서 output이 아닌 다른 형태의 반환값을 가지는 것도 어색한 것 같습니다.

추가로 이름도 전체적으로 뷰의 관점이 아닌, 로직 중심으로 네이밍이 된 것 같아요.
https://github.com/kickstarter/ios-oss/blob/main/Library/ViewModels/LoginViewModel.swift
요기서 input과 output을 어떻게 정의하고 있는지, 이름은 어떻게 지어졌는지 참고해보시면 좋을 것 같습니다 🙂

TTOzzi commented on ProjectManager/ProjectManager/ViewModel/MainViewModel.swift:
```diff
@@ -7,43 +7,88 @@
+    init(storage: Storegeable) {
+        self.storage = storage
+        
+        todoList = storage.list
+            .map{ $0.filter { $0.type == .todo }}
+            .map{ $0.sorted(by: { $0.deadline < $1.deadline}) }
+            .asDriver(onErrorJustReturn: [])
+        
+        doingList = storage.list
+            .map{ $0.filter { $0.type == .doing }}
+            .map{ $0.sorted(by: { $0.deadline < $1.deadline}) }
+            .asDriver(onErrorJustReturn: [])
+        
+        doneList = storage.list
+            .map{ $0.filter { $0.type == .done }}
+            .map{ $0.sorted(by: { $0.deadline < $1.deadline}) }
+            .asDriver(onErrorJustReturn: [])
```
순회 로직이 한 번씩 더 늘었군요 😢

doogie97 replied:
어떤 말씀을 하시는지 잘 이해하지 못했습니다
조금 더 상세하게 말씀해주시면 감사하겠습니다

TTOzzi replied:
지난 PR에서 남겼던 효율성 관련 코멘트의 연장선인데요.
현재 todoList에만 리스트 아이템을 생성해도, `storage.list`는 값을 방출하고 새로 추가된 정렬 로직을 포함해 총 6번의 순회 로직을 돌게 됩니다.
`filter`까진 줄이기 어렵더라도 rx의 `distinctUntilChanged` 연산자를 잘 활용하면 불필요한 정렬 로직을 줄일 수 있을 것 같아요.
크게 체감하긴 어렵긴 하지만, 구현할 때 항상 효율성에 대해 생각하시는 게 좋습니다. 사소한 게 모여서 커지기도 하거든요 🙂

TTOzzi commented on ProjectManager/ProjectManager/ViewModel/MainViewModel.swift:
```diff
@@ -7,43 +7,88 @@
-            .map{ $0.filter { $0.type == type }}
-        return list
+    func peekList(index: Int, type: ListType, completion: @escaping ((ListItem) -> Void)) {
+        switch type {
+        case .todo:
+            todoList.drive(onNext: {
+                completion($0[index])
+            })
+            .dispose()
+        case .doing:
+            doingList.drive(onNext: {
+                completion($0[index])
+            })
+            .dispose()
+        case .done:
+            doneList.drive(onNext: {
+                completion($0[index])
+            })
+            .dispose()
+        }
```
으음 뷰에서 바라봐야 할 output을 뷰모델 내부에서도 바인딩하고 있는데요 🤔 
뷰모델은 데이터의 원천인 `storage`을 알고 있는데, output을 통해 값을 받아와야 할 이유가 있을까요?

위의 input output 흐름에 대한 코멘트를 이어서 작성하자면,
여기선 `completion` 이나 output에 대한 바인딩 없이 데이터의 원천인 `storage` 에서 index 로 값을 찾고 `presentEditView` output을 통해 찾은 값을 넘겨줘야 할 것 같아요.

TTOzzi commented on ProjectManager/ProjectManager/ViewModel/MainViewModel.swift:
```diff
@@ -7,43 +7,88 @@
-        
-        listObservable.accept((todoDummy + doingDummy + doneDummy)
-            .sorted(by: { $0.deadline < $1.deadline}))
+    func deleteList(index: Int, type: ListType) {
+        peekList(index: index, type: type) {
+            self.storage.deleteList(listItem: $0)
+        }
+    }
+    
+    func changeListType(index: Int, type: ListType, to: ListType) {
+        peekList(index: index, type: type) {
+            var listItem = $0
+            listItem.type = to
+            self.storage.updateList(listItem: listItem)
+        }
     }
```
여긴 input이 내부에서 한 번 더 input을 부르는 구조네요 🤔 
마찬가지로 output에서 index에 해당하는 값을 찾는 게 아닌, 데이터의 원천을 관리하는 `storage`를 통해 찾아오고 로직을 수행해야 할 것 같습니다.
로직을 완료한 뒤 뷰에 따로 전달해야 할 업데이트 사항이 있다면 output을 만들어줘야 할 것 같구요.

TTOzzi commented on ProjectManager/ProjectManager/ViewModel/MainViewModel.swift:
```diff
@@ -7,43 +7,88 @@
+    func changeListType(index: Int, type: ListType, to: ListType) {
+        peekList(index: index, type: type) {
+            var listItem = $0
+            listItem.type = to
```
사소하지만 `to` 보단 argumentLabel을 활용해 `to destination` 과 같이 메서드 내부에선 의미있는 이름으로 사용할 수 있도록 해주면 더 좋을 것 같아요.

doogie97 replied:
메서드 내부에서는 의미 있는 이름! 기억하겠습니다

TTOzzi reviewed (changes requested):
안녕하세요 두기!
늦어서 죄송합니다.. 🙏 시작은 어제 했으나.. 제가 리뷰하는데에 시간이 좀 오래걸려서 시간을 훌쩍 넘겨버렸네요 😢 

### input/ output

> input은 viewmodel이 이벤트를 받아 해야할 행동
output은 viewmodel이 view로 전달하는 데이터
정도로 이해를 했는데 알맞게 구현한 것이 맞을까요?

명확한 가이드라인이 없어서 약간의 차이가 있을 수 있지만,
input은 뷰에서 발생할 수 있는 이벤트의 개념으로 봐야 할 것 같고, input을 전달하는 수단이 현재 구현하신 구조에선 함수 호출이 되겠네요.
"이벤트를 받아 해야할 행동"은 input으로 정의된 메서드의 내부 구현이 되겠고, output은 잘 이해하신 것 같습니다 🙂 
다만, 실제 구현에선 어색한 부분이 많이 보이는 것 같아요. 자세한 내용은 코멘트로 남겨두었습니다 🙇 

### cell의 index만 받아 viewModel에서 해당 list를 반환 하기

> 위와 같은 로직을 통해 cell을 반환하는 기능을 구현해 보았는데 completion으로 이렇게 반환하는게 또치가 원하신 모습과 얼마나 차이가 있을지 여쭤보고 싶습니다

요건 코멘트로 남겨뒀어요 🙇 

> 그리고 각각의 output이 driver로 구현되어 있는데 driver는 원래 filter나 map 사용이 안되는건지 여쭤보고 싶습니다(일단 반응을 안해 drive한 뒤에 onNext로 넘겨주도록 구현을 했습니다)

오잉 driver도 operator 사용 가능합니다.
어떤 부분에서 반응을 하지 않던가요? 작성하신 코드가 있다면 더 자세히 확인해드릴 수 있을 것 같아요 🤔 

> drive를 하게 되면 무한 반복되는 현상이 있어서 원하는 일을 한 뒤에 바로 dispose를 해주도록 했는데 이렇게 된다면 문제가 있을까요?(제가 예상하기로는 onNext이후 작업이 오래걸린다면 작업이 끝나기 전에 dispose될 것 같다 라는 생각을 하긴 했는데 이걸 확인 할 방법이 없어 여쭤보게 되었습니다!)

`peekList`메서드는 클릭 이벤트가 발생할 때 마다 호출되는데, driver의 바인딩을 disposeBag에 저장하면 저장한 바인딩이 계속 남아있어 말씀하신대로 로직이 여러번 수행되는 문제가 생길 거예요.
지금 구현하신 코드도 `dispose`는 호출 즉시 바인딩을 폐기해버리기에 예상하셨듯이 시간이 걸리는 로직이었다면 아무 동작도 하지 않았을 것 같습니다.
음.. 지금 코드에선 `delay` operator를 붙여서 확인해볼 순 있겠네요.
일반적인 구현은 아니라 수정이 필요해보여요. 자세한 내용은 코멘트로 남겨뒀는데, 이해가 어려운 부분이 있다면 말씀해주세요 🙇 

고생하셨습니다!

doogie97 committed:
✏️chore : podfile 수정

doogie97 committed:
✏️chore : 접근제어 설정

doogie97 committed:
♻️refactor : dummyList -> list로 변경

doogie97 committed:
♻️refactor : list삭제를 위해 전체 순환이 아닌 특정 index를 찾아 삭제하도록 수정

doogie97 committed:
♻️refactor : alert title과 listType을 Direction으로 묶어줌

doogie97 committed:
✏️chore : 폴더 트리 변경

doogie97 committed:
Merge remote-tracking branch 'origin/STEP2-2' into ic_5_doogie

doogie97 committed:
♻️refactor : Add,EditVC에 있는 listItem을 ViewModel로 이동

doogie97 committed:
💎feat : DetailViewModel의 in/output 정의

doogie97 committed:
♻️refactor : view가 list를 전달하는 것이 아닌 각 요소만 전달하도록 수정

각 뷰의 요소를 받은 viewModel이 자신이 가지고 있는 list를 수정 한 뒤 Done 버튼을 눌렀을 때 자신이 가지고 있는 list로 업데이트 할 수 있도록 구현

doogie97 committed:
♻️refactor : DetailViewModel을 각각 add와 edit으로 분리

doogie97 committed:
⚰️del : 쓸모없는 코드 삭제

doogie97 committed:
♻️refactor : protocol 네이밍 수정

doogie97 committed:
♻️refactor : 버튼 title에 따른 분기 -> viewmodel 상태값으로 분기

doogie97 committed:
✏️chore : 오타 수정

doogie97 committed:
⚰️del : 공백 제거

doogie97 committed:
💎feat : viewmodel의 편집가능여부 상태값을 view와 바인딩

doogie97 committed:
⚰️del : 쓸모없는 코드 삭제

doogie97 committed:
♻️refactor : 메서드 네이밍 수정

doogie97 committed:
♻️refactor : viewModel에서 count를 받아올 수 있도록 수정

doogie97 committed:
♻️refactor : 모호한 네이밍 수정

doogie97 committed:
♻️refactor : 내부 파라미터 추가

doogie97 committed:
♻️refactor : storage의 list를 분리해 Index를 통해 처리 하도록 수정

doogie97 committed:
♻️refactor : EditVC 이동시 받는 파라미터 수정

index와 listType을 통해 storage에서 직접 가져오도록 수정

doogie97 committed:
⚰️del : 쓸모없는 코드 삭제

doogie97 committed:
⚰️del : 쓸모없는 코드 삭제

doogie97 committed:
♻️refactor : selectList -> selectItem 으로 네이밍 수정

doogie97 committed:
♻️refactor : 추상화 중 누락된 부분 수정

doogie97 committed:
♻️refactor : 뷰에서 이벤트 발생시 in으로 받고 적절한 Out을 주도록 수정

doogie97 committed:
🔨fix : addview 진입시에도 edit으로 뜨는 현상 수정

doogie97 committed:
♻️refactor : 뷰에서 이벤트 발생시 in으로 받고 적절한 Out을 주도록 수정

doogie97 committed:
♻️refactor : 뷰에서 이벤트 발생시 in으로 받고 적절한 Out을 주도록 수정

+ editview로 index만 넘겨줘서 detailview에서 item 찾기 -> 미리 찾아서 item을 넘기도록 수정

doogie97 committed:
♻️refactor : isOverDeadline 위치 이동

doogie97 committed:
⚰️del : 쓸모없는 코드 삭제
